Fix non-reentrant hooks based checkpointing

rohan-varma opened the pull request:
Stack from [ghstack](https://github.com/ezyang/ghstack) (oldest at bottom):
* #78854
* __->__ #78752

Fixes the non-reentrant hooks based checkpointing to actually save memory. The issue was that `storage` was a list of autograd saved tensors and we weren't clearing this list out as tensors were accessed, so all activations would remain in memory. Now at the end of the layer's backwards pass, activations will be discarded as expected. 

Adding unittests to ensure:
- Memory savings for a basic model compared to no checkpointing
- Same or better memory savings when compared with the reentrant autograd based hooks checkpointing

Also, this means we can enable non-reentrant based checkpointing in `CheckpointWrapper`, will also add unittests for that.

rohan-varma committed:
[WIP] Fix non-reentrant hooks based checkpointing

[ghstack-poisoned]

albanD commented on torch/utils/checkpoint.py:
```diff
@@ -367,7 +371,7 @@ def inner_unpack(packed):
                         _unused = function(*args)
 
-        return storage[x]
+        return storage.pop(x)
```
We should properly check if the value is there or if it was already removed here.

rohan-varma replied:
@albanD to be clear, this function should always return a tensor, so if it is not, we should raise an error?

albanD replied:
If you can't return the original Tensor, then yes it should raise an error.
That being said, we should try to always make this function work.

rohan-varma replied:
Sure, I think the case that you mentioned works, are there any other edge cases you can think of that we'd like to test?

rohan-varma replied:
yeah, the forward function is re-evaluated when we call `ctx.saved_tensors` again. 

The content of `storage` is not overwritten, because previous contents were removed in the last access, so they are newly added to the dict.

rohan-varma replied:
Figured out how to do above and it is now tested.

albanD reviewed (commented):
Approach sounds ok. We should add a few more checks to make sure users don't end up with weird errors if they do things like this in a custom Function:
```
inp = ctx.saved_tensors[0]
out = ctx.saved_tensors[1]
```
And provide guidance on what to do to avoid getting an error in this case.

rohan-varma committed:
Update on "[WIP] Fix non-reentrant hooks based checkpointing"


Fixes the non-reentrant hooks based checkpointing to actually save memory. The issue was that `storage` was a list of autograd saved tensors and we weren't clearing this list out as tensors were accessed, so all activations would remain in memory. Now at the end of the layer's backwards pass, activations will be discarded as expected. 

Adding unittests to ensure:
- Memory savings for a basic model compared to no checkpointing
- Same or better memory savings when compared with the reentrant autograd based hooks checkpointing

Also, this means we can enable non-reentrant based checkpointing in `CheckpointWrapper`, will also add unittests for that.

[ghstack-poisoned]

rohan-varma committed:
Update on "[WIP] Fix non-reentrant hooks based checkpointing"


Fixes the non-reentrant hooks based checkpointing to actually save memory. The issue was that `storage` was a list of autograd saved tensors and we weren't clearing this list out as tensors were accessed, so all activations would remain in memory. Now at the end of the layer's backwards pass, activations will be discarded as expected. 

Adding unittests to ensure:
- Memory savings for a basic model compared to no checkpointing
- Same or better memory savings when compared with the reentrant autograd based hooks checkpointing

Also, this means we can enable non-reentrant based checkpointing in `CheckpointWrapper`, will also add unittests for that.

[ghstack-poisoned]

rohan-varma committed:
Update on "[WIP] Fix non-reentrant hooks based checkpointing"


Fixes the non-reentrant hooks based checkpointing to actually save memory. The issue was that `storage` was a list of autograd saved tensors and we weren't clearing this list out as tensors were accessed, so all activations would remain in memory. Now at the end of the layer's backwards pass, activations will be discarded as expected. 

Adding unittests to ensure:
- Memory savings for a basic model compared to no checkpointing
- Same or better memory savings when compared with the reentrant autograd based hooks checkpointing

Also, this means we can enable non-reentrant based checkpointing in `CheckpointWrapper`, will also add unittests for that.

[ghstack-poisoned]

rohan-varma committed:
Update on "[WIP] Fix non-reentrant hooks based checkpointing"


Fixes the non-reentrant hooks based checkpointing to actually save memory. The issue was that `storage` was a list of autograd saved tensors and we weren't clearing this list out as tensors were accessed, so all activations would remain in memory. Now at the end of the layer's backwards pass, activations will be discarded as expected. 

Adding unittests to ensure:
- Memory savings for a basic model compared to no checkpointing
- Same or better memory savings when compared with the reentrant autograd based hooks checkpointing

Also, this means we can enable non-reentrant based checkpointing in `CheckpointWrapper`, will also add unittests for that.

[ghstack-poisoned]

rohan-varma committed:
Update on "[WIP] Fix non-reentrant hooks based checkpointing"


Fixes the non-reentrant hooks based checkpointing to actually save memory. The issue was that `storage` was a list of autograd saved tensors and we weren't clearing this list out as tensors were accessed, so all activations would remain in memory. Now at the end of the layer's backwards pass, activations will be discarded as expected. 

Adding unittests to ensure:
- Memory savings for a basic model compared to no checkpointing
- Same or better memory savings when compared with the reentrant autograd based hooks checkpointing

Also, this means we can enable non-reentrant based checkpointing in `CheckpointWrapper`, will also add unittests for that.

[ghstack-poisoned]

rohan-varma commented:
> Approach sounds ok. We should add a few more checks to make sure users don't end up with weird errors if they do things like this in a custom Function:
> 
> ```
> inp = ctx.saved_tensors[0]
> out = ctx.saved_tensors[1]
> ```
> 
> And provide guidance on what to do to avoid getting an error in this case.

This appears to work, i.e. based on a unittest we are able to access the saved tensors > 1x.

rohan-varma committed:
Update on "[WIP] Fix non-reentrant hooks based checkpointing"


Fixes the non-reentrant hooks based checkpointing to actually save memory. The issue was that `storage` was a list of autograd saved tensors and we weren't clearing this list out as tensors were accessed, so all activations would remain in memory. Now at the end of the layer's backwards pass, activations will be discarded as expected. 

Adding unittests to ensure:
- Memory savings for a basic model compared to no checkpointing
- Same or better memory savings when compared with the reentrant autograd based hooks checkpointing

Also, this means we can enable non-reentrant based checkpointing in `CheckpointWrapper`, will also add unittests for that.

[ghstack-poisoned]

rohan-varma committed:
Update on "[WIP] Fix non-reentrant hooks based checkpointing"


Fixes the non-reentrant hooks based checkpointing to actually save memory. The issue was that `storage` was a list of autograd saved tensors and we weren't clearing this list out as tensors were accessed, so all activations would remain in memory. Now at the end of the layer's backwards pass, activations will be discarded as expected. 

Adding unittests to ensure:
- Memory savings for a basic model compared to no checkpointing
- Same or better memory savings when compared with the reentrant autograd based hooks checkpointing

Also, this means we can enable non-reentrant based checkpointing in `CheckpointWrapper`, will also add unittests for that.

[ghstack-poisoned]

albanD commented on test/test_autograd.py:
```diff
@@ -4436,6 +4436,82 @@ def test_checkpointing(self):
+                ctx.save_for_backward(x, y, z, w, out)
+                return out
+
+            def backward(ctx, grad_out):
```
Should be static method

albanD commented on torch/utils/checkpoint.py:
```diff
@@ -367,7 +371,9 @@ def inner_unpack(packed):
 
-        return storage[x]
+        if x not in storage:
+            raise RuntimeError("Attempted to unpack tensor that has not been packed, please report a bug to PyTorch")
```
I'm not sure that sentence makes much sense from the end user point of view? I think "checkpoint" is ok but "packed" is a bit weird.

rohan-varma replied:
Sure, how about:

`Attempted to retrieve tensor that has not been checkpointed, please report a bug to pytorch`?

albanD replied:
I think we should mention that this is about retrieving it the tensor twice?
`Retrieving a Tensor saved by autograd multiple times while checkpointing is not currently supported. Please open an issue with details on your use case so that we can prioritize adding this`.

rohan-varma commented on test/test_autograd.py:
```diff
@@ -4436,6 +4436,82 @@ def test_checkpointing(self):
+        model_reentrant_checkpoint = MyModel(8, use_checkpoint=True, use_reentrant=True).cuda()
+        model_no_reentrant_checkpoint = MyModel(8, use_checkpoint=True, use_reentrant=False).cuda()
+
+        # Large batch size to ensure activation sizes >> gradient sizes, so activation checkpointing
```
Sure, as in setting p.grad before forward pass?

rohan-varma replied:
fixed

rohan-varma committed:
Update on "[WIP] Fix non-reentrant hooks based checkpointing"


Fixes the non-reentrant hooks based checkpointing to actually save memory. The issue was that `storage` was a list of autograd saved tensors and we weren't clearing this list out as tensors were accessed, so all activations would remain in memory. Now at the end of the layer's backwards pass, activations will be discarded as expected. 

Adding unittests to ensure:
- Memory savings for a basic model compared to no checkpointing
- Same or better memory savings when compared with the reentrant autograd based hooks checkpointing

Also, this means we can enable non-reentrant based checkpointing in `CheckpointWrapper`, will also add unittests for that.

[ghstack-poisoned]

rohan-varma commented on test/test_autograd.py:
```diff
@@ -4436,6 +4436,82 @@ def test_checkpointing(self):
+                self.layers = nn.ModuleList()
+                for i in range(self.n):
+                    layer = nn.Sequential(
+                        nn.Linear(256, 256), nn.Linear(256, 256), nn.Linear(256, 256)
```
just tested and it runs in ~2s on on AWS cluster GPU instance.

rohan-varma committed:
Update on "Fix non-reentrant hooks based checkpointing"


Fixes the non-reentrant hooks based checkpointing to actually save memory. The issue was that `storage` was a list of autograd saved tensors and we weren't clearing this list out as tensors were accessed, so all activations would remain in memory. Now at the end of the layer's backwards pass, activations will be discarded as expected. 

Adding unittests to ensure:
- Memory savings for a basic model compared to no checkpointing
- Same or better memory savings when compared with the reentrant autograd based hooks checkpointing

Also, this means we can enable non-reentrant based checkpointing in `CheckpointWrapper`, will also add unittests for that.

[ghstack-poisoned]

albanD commented on test/test_autograd.py:
```diff
@@ -4436,6 +4436,94 @@ def test_checkpointing(self):
+        loss.backward()
+        mem_no_reentrant_checkpoint = torch.cuda.max_memory_allocated()
+
+        print(mem_no_checkpoint, mem_reentrant_checkpoint, mem_no_reentrant_checkpoint)
```
print

albanD commented on test/test_autograd.py:
```diff
@@ -4436,6 +4436,94 @@ def test_checkpointing(self):
+
+        print(mem_no_checkpoint, mem_reentrant_checkpoint, mem_no_reentrant_checkpoint)
+        self.assertTrue(mem_reentrant_checkpoint < mem_no_checkpoint)
+        self.assertTrue(mem_no_reentrant_checkpoint <= mem_reentrant_checkpoint)
```
Why would this one be "or equal"?

albanD replied:
I'm not sure we can really predict how the reentrant vs non-reentrant will behave between each other.
`mem_no_reentrant_checkpoint < mem_no_checkpoint` may be a better second assert/

rohan-varma replied:
Sounds good, changed it to this

albanD commented on test/test_autograd.py:
```diff
@@ -4436,6 +4436,94 @@ def test_checkpointing(self):
+            def backward(ctx, grad_out):
+                x, y, z, w, out = ctx.saved_tensors
+                x_2, y_2, z_2, w_2, out_2 = ctx.saved_tensors
+                self.assertEqual(x, x_2)
```
Should this be `assertIs` ?

rohan-varma replied:
I think x, y, z can be assertIs, w and out cannot be, because they are recomputed in forward pass

rohan-varma commented on torch/utils/checkpoint.py:
```diff
@@ -367,7 +371,13 @@ def inner_unpack(packed):
 
-        return storage[x]
+        if x not in storage:
+            raise RuntimeError(
```
Agreed, although I'm not sure how to reproduce this as the forward is re-run every time we access `saved_tensors`. Do you have any suggestions for a potential use case that would hit this issue?

albanD replied:
You only recompute if `len(storage)` is 0 though. So I guess that if you add more ops before your custom Function in the current example that should trigger it?

rohan-varma replied:
This is currently the usage of the custom function:

```
x = torch.tensor(1., requires_grad=True)
        y = torch.tensor(2., requires_grad=True)
        z = torch.tensor(3., requires_grad=True)
        out = checkpoint(MyFunc.apply, x, y, z, use_reentrant=False)
        out.sum().backward()
```

Maybe I'm missing something, but how would adding more ops before the custom function call trigger this? That should entirely be out of the checkpoint context?

rohan-varma replied:
Discussed offline - we'll trigger the error by accessing _saved_result

rohan-varma commented on torch/utils/checkpoint.py:
```diff
@@ -343,10 +343,14 @@ def pack(x):
         return counter - 1
 
     def unpack(x):
+        nonlocal storage
```
removed

rohan-varma committed:
Update on "Fix non-reentrant hooks based checkpointing"


Fixes the non-reentrant hooks based checkpointing to actually save memory. The issue was that `storage` was a list of autograd saved tensors and we weren't clearing this list out as tensors were accessed, so all activations would remain in memory. Now at the end of the layer's backwards pass, activations will be discarded as expected. 

Adding unittests to ensure:
- Memory savings for a basic model compared to no checkpointing
- Same or better memory savings when compared with the reentrant autograd based hooks checkpointing

Also, this means we can enable non-reentrant based checkpointing in `CheckpointWrapper`, will also add unittests for that.

[ghstack-poisoned]

albanD commented on test/test_autograd.py:
```diff
@@ -4436,6 +4436,95 @@ def test_checkpointing(self):
+        x = x * y * z
+        y = y * y * z
+        z = z * z
+        out = checkpoint(MyFunc.apply, x, y, z, use_reentrant=False)
```
You need to include the previous computations in the checkpoint for the error to be triggered!

rohan-varma replied:
Discussed offline - we'll trigger the error by accessing `_saved_result`

rohan-varma committed:
Update on "Fix non-reentrant hooks based checkpointing"


Fixes the non-reentrant hooks based checkpointing to actually save memory. The issue was that `storage` was a list of autograd saved tensors and we weren't clearing this list out as tensors were accessed, so all activations would remain in memory. Now at the end of the layer's backwards pass, activations will be discarded as expected. 

Adding unittests to ensure:
- Memory savings for a basic model compared to no checkpointing
- Same or better memory savings when compared with the reentrant autograd based hooks checkpointing

Also, this means we can enable non-reentrant based checkpointing in `CheckpointWrapper`, will also add unittests for that.

[ghstack-poisoned]

rohan-varma committed:
Update on "Fix non-reentrant hooks based checkpointing"


Fixes the non-reentrant hooks based checkpointing to actually save memory. The issue was that `storage` was a list of autograd saved tensors and we weren't clearing this list out as tensors were accessed, so all activations would remain in memory. Now at the end of the layer's backwards pass, activations will be discarded as expected. 

Adding unittests to ensure:
- Memory savings for a basic model compared to no checkpointing
- Same or better memory savings when compared with the reentrant autograd based hooks checkpointing

Also, this means we can enable non-reentrant based checkpointing in `CheckpointWrapper`, will also add unittests for that.

[ghstack-poisoned]

rohan-varma committed:
Update on "Fix non-reentrant hooks based checkpointing"


Fixes the non-reentrant hooks based checkpointing to actually save memory. The issue was that `storage` was a list of autograd saved tensors and we weren't clearing this list out as tensors were accessed, so all activations would remain in memory. Now at the end of the layer's backwards pass, activations will be discarded as expected. 

Adding unittests to ensure:
- Memory savings for a basic model compared to no checkpointing
- Same or better memory savings when compared with the reentrant autograd based hooks checkpointing

Also, this means we can enable non-reentrant based checkpointing in `CheckpointWrapper`, will also add unittests for that.

[ghstack-poisoned]

rohan-varma committed:
Update on "Fix non-reentrant hooks based checkpointing"


Fixes the non-reentrant hooks based checkpointing to actually save memory. The issue was that `storage` was a list of autograd saved tensors and we weren't clearing this list out as tensors were accessed, so all activations would remain in memory. Now at the end of the layer's backwards pass, activations will be discarded as expected. 

Adding unittests to ensure:
- Memory savings for a basic model compared to no checkpointing
- Same or better memory savings when compared with the reentrant autograd based hooks checkpointing

Also, this means we can enable non-reentrant based checkpointing in `CheckpointWrapper`, will also add unittests for that.

[ghstack-poisoned]

albanD commented on test/test_autograd.py:
```diff
@@ -4436,6 +4436,119 @@ def test_checkpointing(self):
+            @staticmethod
+            def backward(ctx, grad_out):
+                x, y, z, w, out = ctx.saved_tensors
+                # This access is fine as entire saved_tensors is
```
I don't think this is accurate! If any op is checkpointed with this custom function (which will necessarily happen as checkpointing just this is useless), then this won't work. I think we want to change this example.

rohan-varma replied:
Fixed to error properly as discussed offline.

albanD reviewed (commented):
All looks good but the custom Function test.

rohan-varma committed:
Update on "Fix non-reentrant hooks based checkpointing"


Fixes the non-reentrant hooks based checkpointing to actually save memory. The issue was that `storage` was a list of autograd saved tensors and we weren't clearing this list out as tensors were accessed, so all activations would remain in memory. Now at the end of the layer's backwards pass, activations will be discarded as expected. 

Adding unittests to ensure:
- Memory savings for a basic model compared to no checkpointing
- Same or better memory savings when compared with the reentrant autograd based hooks checkpointing

Also, this means we can enable non-reentrant based checkpointing in `CheckpointWrapper`, will also add unittests for that.

[ghstack-poisoned]

rohan-varma committed:
Update on "Fix non-reentrant hooks based checkpointing"


Fixes the non-reentrant hooks based checkpointing to actually save memory. The issue was that `storage` was a list of autograd saved tensors and we weren't clearing this list out as tensors were accessed, so all activations would remain in memory. Now at the end of the layer's backwards pass, activations will be discarded as expected. 

Adding unittests to ensure:
- Memory savings for a basic model compared to no checkpointing
- Same or better memory savings when compared with the reentrant autograd based hooks checkpointing

Also, this means we can enable non-reentrant based checkpointing in `CheckpointWrapper`, will also add unittests for that.

[ghstack-poisoned]

albanD reviewed (approved):
SGTM

awgu commented on torch/utils/checkpoint.py:
```diff
@@ -1,6 +1,6 @@
 import torch
 import warnings
-from typing import Any, Iterable, List, Tuple, Union
+from typing import Any, Iterable, List, Tuple, Union, Dict
```
nit: sort

awgu commented on torch/utils/checkpoint.py:
```diff
@@ -332,7 +332,7 @@ def _checkpoint_without_reentrant(function, preserve_rng_state=True, *args):
             fwd_gpu_devices, fwd_gpu_states = get_device_states(*args)
 
-    storage: List[Union[torch.Tensor, None]] = []
+    storage: Dict[int, Union[torch.Tensor, None]] = {}
```
nit: `Union[torch.Tensor, None]` -> `Optional[torch.Tensor]`

awgu commented on test/distributed/fsdp/test_checkpoint_wrapper.py:
```diff
@@ -40,6 +48,61 @@ def test_load_activation_checkpointed_module(self):
+        equivalent memory usage wise.
+        """
+        class Model(nn.Module):
+            def __init__(self, n: int, use_cp: bool, use_wrapper: bool = False, use_reentrant=True):
```
nit: Can we clean this up to be consistent?
For example:
```suggestion
            def __init__(self, n: int, use_cp: bool, use_wrapper: bool = False, use_reentrant: bool = True):
```

awgu commented on test/distributed/fsdp/test_checkpoint_wrapper.py:
```diff
@@ -40,6 +48,61 @@ def test_load_activation_checkpointed_module(self):
+                return x
+
+        def test(use_checkpointing, use_wrapper, use_reentrant):
+            a = Model(8, use_checkpointing, use_wrapper=use_wrapper, use_reentrant=False).cuda()
```
Is the argument `use_reentrant` to `test()` used?

awgu reviewed (commented):
I may be misreading, but check `test_checkpoint_wrapper_parity()`. It looks like `use_reentrant` is not used.

rohan-varma committed:
Update on "Fix non-reentrant hooks based checkpointing"


Fixes the non-reentrant hooks based checkpointing to actually save memory. The issue was that `storage` was a list of autograd saved tensors and we weren't clearing this list out as tensors were accessed, so all activations would remain in memory. Now at the end of the layer's backwards pass, activations will be discarded as expected. 

Adding unittests to ensure:
- Memory savings for a basic model compared to no checkpointing
- Same or better memory savings when compared with the reentrant autograd based hooks checkpointing

Also, this means we can enable non-reentrant based checkpointing in `CheckpointWrapper`, will also add unittests for that.

[ghstack-poisoned]

rohan-varma committed:
Update on "Fix non-reentrant hooks based checkpointing"


Fixes the non-reentrant hooks based checkpointing to actually save memory. The issue was that `storage` was a list of autograd saved tensors and we weren't clearing this list out as tensors were accessed, so all activations would remain in memory. Now at the end of the layer's backwards pass, activations will be discarded as expected. 

Adding unittests to ensure:
- Memory savings for a basic model compared to no checkpointing
- Same or better memory savings when compared with the reentrant autograd based hooks checkpointing

Also, this means we can enable non-reentrant based checkpointing in `CheckpointWrapper`, will also add unittests for that.

[ghstack-poisoned]

nyngwang commented:
Awesome!
